File: skills/runbook/scripts/commands/shift_items.py
from __future__ import annotations

import argparse
import sys
from pathlib import Path

import difflib
import re
from dataclasses import dataclass

from commands.normalize import NUMBERED_H3_RE, extract_h3_blocks, parse_sections, section_slice


HEADING_RE = re.compile(r"^(### )(?:[🟢🟡🔴]\s+)?(\d+)(\. .*)$")
ITEM_TOKEN_RE = re.compile(r"item-(\d+)(-execution-record|-acceptance-record)?")


@dataclass(frozen=True)
class Step:
    number: int
    label: str
    heading_idx: int


def load_numbered_steps(lines: list[str], section_name: str) -> tuple[tuple[int, int], list[Step]]:
    h2_sections = parse_sections(lines, 2)
    section = section_slice(h2_sections, section_name, len(lines))
    if section is None:
        raise ValueError(f"missing `## {section_name}` section")

    start, end = section
    steps: list[Step] = []
    for heading_idx, title, _, _ in extract_h3_blocks(lines, start + 1, end):
        match = NUMBERED_H3_RE.match(title)
        if match is None:
            raise ValueError(f"`## {section_name}` contains a non-numbered step heading: {title}")
        steps.append(Step(number=int(match.group(1)), label=match.group(2), heading_idx=heading_idx))
    if not steps:
        raise ValueError(f"`## {section_name}` does not contain any numbered items")
    return section, steps


def ensure_step_alignment(plan_steps: list[Step], record_steps: list[Step]) -> None:
    plan_shape = [(step.number, step.label) for step in plan_steps]
    record_shape = [(step.number, step.label) for step in record_steps]
    if plan_shape != record_shape:
        raise ValueError("`## 执行计划` and `## 执行记录` are not aligned; renumbering aborted")


def build_mapping(steps: list[Step], start: int, shift: int) -> dict[int, int]:
    if start < 1:
        raise ValueError("`--start` must be >= 1")
    if shift < 1:
        raise ValueError("`--shift` must be >= 1")

    existing_numbers = {step.number for step in steps}
    if start not in existing_numbers:
        available = ", ".join(str(step.number) for step in steps)
        raise ValueError(f"`--start {start}` does not match an existing item; available numbers: {available}")

    return {step.number: step.number + shift for step in steps if step.number >= start}


def replace_heading(line: str, mapping: dict[int, int]) -> str:
    newline = "\n" if line.endswith("\n") else ""
    body = line[:-1] if newline else line
    match = HEADING_RE.match(body)
    if match is None:
        return line

    old_number = int(match.group(2))
    new_number = mapping.get(old_number)
    if new_number is None:
        return line
    traffic_light = ""
    title_body = body.removeprefix(match.group(1))
    if title_body.startswith(("🟢 ", "🟡 ", "🔴 ")):
        traffic_light = title_body[:2]
    return f"{match.group(1)}{traffic_light}{new_number}{match.group(3)}{newline}"


def replace_item_tokens(line: str, mapping: dict[int, int]) -> str:
    def repl(match: re.Match[str]) -> str:
        old_number = int(match.group(1))
        suffix = match.group(2) or ""
        new_number = mapping.get(old_number)
        if new_number is None:
            return match.group(0)
        return f"item-{new_number}{suffix}"

    return ITEM_TOKEN_RE.sub(repl, line)


def rewrite_section(lines: list[str], section: tuple[int, int], mapping: dict[int, int]) -> None:
    start, end = section
    for idx in range(start + 1, end):
        line = lines[idx]
        line = replace_heading(line, mapping)
        line = replace_item_tokens(line, mapping)
        lines[idx] = line


def shift_runbook_items(text: str, start: int, shift: int) -> tuple[str, dict[int, int]]:
    lines = text.splitlines(keepends=True)
    if not lines:
        raise ValueError("runbook is empty")

    plain_lines = [line.rstrip("\n") for line in lines]
    plan_section, plan_steps = load_numbered_steps(plain_lines, "执行计划")
    record_section, record_steps = load_numbered_steps(plain_lines, "执行记录")
    ensure_step_alignment(plan_steps, record_steps)

    mapping = build_mapping(plan_steps, start=start, shift=shift)
    updated_lines = list(lines)
    rewrite_section(updated_lines, plan_section, mapping)
    rewrite_section(updated_lines, record_section, mapping)
    return "".join(updated_lines), mapping


def render_diff(path: Path, before: str, after: str) -> str:
    return "".join(
        difflib.unified_diff(
            before.splitlines(keepends=True),
            after.splitlines(keepends=True),
            fromfile=str(path),
            tofile=str(path),
        )
    )


def register(subparsers: argparse._SubParsersAction[argparse.ArgumentParser]) -> None:
    parser = subparsers.add_parser(
        "shift-items",
        help="Shift numbered runbook items to open insertion slots",
        description=(
            "Shift numbered items in `## 执行计划` and `## 执行记录` "
            "to open insertion slots in the middle of a runbook."
        ),
    )
    parser.add_argument("runbook", type=Path, help="Path to the runbook markdown file")
    parser.add_argument("--start", type=int, required=True, help="First existing item number to shift")
    parser.add_argument("--shift", type=int, required=True, help="How many slots to open")
    parser.add_argument(
        "--in-place",
        action="store_true",
        help="Write the rewritten markdown back to the source file",
    )
    parser.set_defaults(handler=handle)


def handle(args: argparse.Namespace) -> int:
    source = args.runbook.read_text(encoding="utf-8")
    try:
        rewritten, mapping = shift_runbook_items(source, start=args.start, shift=args.shift)
    except ValueError as exc:
        print(f"error: {exc}", file=sys.stderr)
        return 1

    if rewritten == source:
        print("No numbering changes were required.", file=sys.stderr)
        return 0

    opened_end = args.start + args.shift - 1
    moved_count = len(mapping)
    moved_summary = f"{min(mapping)}-{max(mapping)} -> {min(mapping.values())}-{max(mapping.values())}"

    if args.in_place:
        args.runbook.write_text(rewritten, encoding="utf-8")
        print(
            f"Shifted {moved_count} item(s) in {args.runbook}; "
            f"opened slots {args.start}-{opened_end}; moved {moved_summary}."
        )
        return 0

    diff = render_diff(args.runbook, source, rewritten)
    if diff:
        sys.stdout.write(diff)
    else:
        print("No numbering changes were required.", file=sys.stderr)
    print(
        f"Preview only: would open slots {args.start}-{opened_end} by shifting {moved_count} item(s).",
        file=sys.stderr,
    )
    return 0

File: skills/runbook/scripts/commands/shift_items.mjs
import path from "node:path";
import { promises as fs } from "node:fs";
import { NUMBERED_H3_RE, extractH3Blocks, parseSections, sectionSlice } from "./normalize.mjs";

const HEADING_RE = /^(### )(?:[🟢🟡🔴]\s+)?(\d+)(\. .*)$/u;
const ITEM_TOKEN_RE = /item-(\d+)(-execution-record|-acceptance-record)?/g;
const TRAFFIC_PREFIXES = ["🟢 ", "🟡 ", "🔴 "];

function loadNumberedSteps(lines, sectionName) {
  const h2Sections = parseSections(lines, 2);
  const section = sectionSlice(h2Sections, sectionName, lines.length);
  if (!section) {
    throw new Error(`missing \`## ${sectionName}\` section`);
  }
  const [start, end] = section;
  const steps = [];
  for (const [headingIdx, title] of extractH3Blocks(lines, start + 1, end)) {
    const match = title.match(NUMBERED_H3_RE);
    if (!match) {
      throw new Error(`\`## ${sectionName}\` contains a non-numbered step heading: ${title}`);
    }
    steps.push({ number: Number(match[1]), label: match[2], headingIdx });
  }
  if (steps.length === 0) {
    throw new Error(`\`## ${sectionName}\` does not contain any numbered items`);
  }
  return [section, steps];
}

function ensureStepAlignment(planSteps, recordSteps) {
  const planShape = planSteps.map((step) => [step.number, step.label]);
  const recordShape = recordSteps.map((step) => [step.number, step.label]);
  if (JSON.stringify(planShape) !== JSON.stringify(recordShape)) {
    throw new Error("`## 执行计划` and `## 执行记录` are not aligned; renumbering aborted");
  }
}

function buildMapping(steps, start, shift) {
  if (start < 1) {
    throw new Error("`--start` must be >= 1");
  }
  if (shift < 1) {
    throw new Error("`--shift` must be >= 1");
  }
  const existingNumbers = new Set(steps.map((step) => step.number));
  if (!existingNumbers.has(start)) {
    throw new Error(`\`--start ${start}\` does not match an existing item; available numbers: ${steps.map((step) => step.number).join(", ")}`);
  }
  const mapping = new Map();
  for (const step of steps) {
    if (step.number >= start) {
      mapping.set(step.number, step.number + shift);
    }
  }
  return mapping;
}

function replaceHeading(line, mapping) {
  const newline = line.endsWith("\n") ? "\n" : "";
  const body = newline ? line.slice(0, -1) : line;
  const match = body.match(HEADING_RE);
  if (!match) {
    return line;
  }
  const oldNumber = Number(match[2]);
  const newNumber = mapping.get(oldNumber);
  if (!newNumber) {
    return line;
  }
  let trafficLight = "";
  const titleBody = body.slice(match[1].length);
  trafficLight = TRAFFIC_PREFIXES.find((prefix) => titleBody.startsWith(prefix)) ?? "";
  return `${match[1]}${trafficLight}${newNumber}${match[3]}${newline}`;
}

function replaceItemTokens(line, mapping) {
  return line.replace(ITEM_TOKEN_RE, (full, number, suffix = "") => {
    const newNumber = mapping.get(Number(number));
    return newNumber ? `item-${newNumber}${suffix}` : full;
  });
}

function rewriteSection(lines, section, mapping, { rewriteHeadings }) {
  const [start, end] = section;
  for (let idx = start + 1; idx < end; idx += 1) {
    let line = rewriteHeadings ? replaceHeading(lines[idx], mapping) : lines[idx];
    line = replaceItemTokens(line, mapping);
    lines[idx] = line;
  }
}

export function shiftRunbookItems(text, start, shift) {
  const lines = text.match(/[^\n]*\n|[^\n]+/g) ?? [];
  if (lines.length === 0) {
    throw new Error("runbook is empty");
  }
  const plainLines = lines.map((line) => line.replace(/\n$/, ""));
  const [planSection, planSteps] = loadNumberedSteps(plainLines, "执行计划");
  const [recordSection, recordSteps] = loadNumberedSteps(plainLines, "执行记录");
  ensureStepAlignment(planSteps, recordSteps);
  const mapping = buildMapping(planSteps, start, shift);
  const updatedLines = [...lines];
  rewriteSection(updatedLines, planSection, mapping, { rewriteHeadings: true });
  rewriteSection(updatedLines, recordSection, mapping, { rewriteHeadings: false });
  return [updatedLines.join(""), Object.fromEntries(mapping)];
}

export function renderDiff(filePath, before, after) {
  const beforeLines = before.split("\n");
  const afterLines = after.split("\n");
  let output = `--- ${filePath}\n+++ ${filePath}\n`;
  const max = Math.max(beforeLines.length, afterLines.length);
  for (let i = 0; i < max; i += 1) {
    if (beforeLines[i] === afterLines[i]) {
      continue;
    }
    if (beforeLines[i] != null) {
      output += `-${beforeLines[i]}\n`;
    }
    if (afterLines[i] != null) {
      output += `+${afterLines[i]}\n`;
    }
  }
  return output;
}

export async function handleShiftItems(args) {
  const filePath = path.resolve(args.runbook);
  const source = await fs.readFile(filePath, "utf8");
  let rewritten;
  let mapping;
  try {
    [rewritten, mapping] = shiftRunbookItems(source, args.start, args.shift);
  } catch (error) {
    console.error(`error: ${error.message}`);
    return 1;
  }
  if (rewritten === source) {
    console.error("No numbering changes were required.");
    return 0;
  }
  const openedEnd = args.start + args.shift - 1;
  const movedKeys = Object.keys(mapping).map(Number);
  const movedSummary = `${Math.min(...movedKeys)}-${Math.max(...movedKeys)} -> ${Math.min(...Object.values(mapping))}-${Math.max(...Object.values(mapping))}`;
  if (args.inPlace) {
    await fs.writeFile(filePath, rewritten, "utf8");
    console.log(`Shifted ${movedKeys.length} item(s) in ${filePath}; opened slots ${args.start}-${openedEnd}; moved ${movedSummary}.`);
    return 0;
  }
  const diff = renderDiff(filePath, source, rewritten);
  if (diff) {
    process.stdout.write(diff);
  } else {
    console.error("No numbering changes were required.");
  }
  console.error(`Preview only: would open slots ${args.start}-${openedEnd} by shifting ${movedKeys.length} item(s).`);
  return 0;
}
